### PR TITLE
merge tree sink cancel delayed_chunk

### PR DESCRIPTION
**Pull request by CheSema**
Motivation and context are here https://github.com/ClickHouse/ClickHouse/pull/66144#issuecomment-2217905961

### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
[Replicated]MergeTreeSink has to properly cancel its delayed_chunk on `onCancel()` method. 

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

#### CI Settings (Only check the boxes if you know what you are doing):
- [ ]  Allow: All Required Checks
- [ ]  Allow: Stateless tests
- [ ]  Allow: Stateful tests
- [ ]  Allow: Integration Tests
- [ ]  Allow: Performance tests
- [ ]  Allow: All Builds
- [ ]  Allow: batch 1, 2 for multi-batch jobs
- [ ]  Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ]  Exclude: Style check
- [ ]  Exclude: Fast test
- [ ]  Exclude: All with ASAN
- [ ]  Exclude: All with TSAN, MSAN, UBSAN, Coverage
- [ ]  Exclude: All with aarch64, release, debug
---
- [ ]  Do not test
- [x]  Woolen Wolfdog
- [ ]  Upload binaries for special builds
- [ ]  Disable merge-commit
- [ ]  Disable CI cache

### CODE REVIEW

**Comment by robot-ch-test-poll1** [comment]
*This is an automated comment for commit e7d6bfa5dc5ee3121525538bb3c9c54c34942fc8 with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/66279/e7d6bfa5dc5ee3121525538bb3c9c54c34942fc8/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Docs check</td><td>Builds and tests the documentation</td><td>✅ success</td></tr>
<tbody>
</table>
</details>

**Comment by CheSema** [comment]
We have a problem. `onCancel()` is called concurrently. There should be synchronisation.

**Comment by CheSema** [comment]
I didn't fully understood how `onCancel` and `onException` should work.
Now I think I do. Here is the explanation:

There is no need to do anything in that methods. Because if you have an `onException` graph execution ether would mute it or cancel the hole graph. If exception is muted that you have to only properly finish your processor as canceled which observed that exception. Other processor might have newer seen any errors, they have to be finished as successful processors. In any case there is an invariant: if processor is finished as successful then method `onFinish` is called. Otherwise there is no `onFinish` method, just destructor.

| relation to exception /       | exception is muted | exception is propagated  |
| :---        |    :----:   |          ---: |
|  trows/observes exception      | `onException` + d-tor   | `onException` + `onCancel` + d-tor.   |
| unaware of exception   | `onFinish` + d-tor        |    `onCancel` + d-tor.    |

So, we could just look at `isCanceled` on d-tor to understand what we should do with the resources: ether cancel them or commit.

Note that `onCancel` is called concurrently with  `onConsume` or `onGenerate` in order to make a chance to stop them earlier. So the synchronisation is required. 
Therefore, it is make cense to ether raise some thread-safe flag or periodically look at `isCanceled` in your `onConsume` or `onGenerate` methods (also in successor: `consume`) to prevent making excess work.

**Comment by CheSema** [comment]
Integration Tests (asan, Old Analyzer) [1/6]
test_broken_ignored_replicated -- flaks still even after some fix https://github.com/ClickHouse/ClickHouse/pull/61932

Stateless Tests (coverage) [2/5]
03030_system_flush_distributed_settings -- flaks in master

Stateless Tests (debug) [1/2]
01632_tinylog_read_write -- assertion in jemalloc https://github.com/ClickHouse/ClickHouse/issues/66193

File: src/Storages/MessageQueueSink.h [hunk 1]
```diff
@@ -33,13 +33,17 @@ class MessageQueueSink : public SinkToStorage
         const String & storage_name_,
         const ContextPtr & context_);
 
+    ~MessageQueueSink() override
+    {
+        onFinish();
```
Comment (CheSema) [inline src/Storages/MessageQueueSink.h:38]: That is not cleat if it could throw an exception. Will revert changes in that class.

**Comment by CheSema** [comment]
Stress test (msan) 
 use-of-uninitialized-value -- https://github.com/ClickHouse/ClickHouse/issues/66525

binary_tidy -- https://github.com/ClickHouse/ClickHouse/pull/66552

File: src/Processors/Executors/CompletedPipelineExecutor.cpp [hunk 2]
```diff
@@ -3,6 +3,7 @@
 #include <QueryPipeline/QueryPipeline.h>
 #include <QueryPipeline/ReadProgressCallback.h>
 #include <Poco/Event.h>
+#include "Common/Logger.h"
```
Comment (antonio2368) [inline src/Processors/Executors/CompletedPipelineExecutor.cpp:6]: Probably added for some deleted logs?

File: src/Storages/MergeTree/MergeTreeDataWriter.cpp [hunk 3]
```diff
@@ -184,6 +186,8 @@ void updateTTL(
 
 void MergeTreeDataWriter::TemporaryPart::cancel()
 {
+    LOG_INFO(getLogger("MergeTreeDataWriter"), "TemporaryPart cancel");
```
Comment (antonio2368) [inline src/Storages/MergeTree/MergeTreeDataWriter.cpp:189]: Debug log?

File: src/Processors/Executors/CompletedPipelineExecutor.cpp (hunk 2 above)
Comment (CheSema) [reply src/Processors/Executors/CompletedPipelineExecutor.cpp]: Yes, I deleted it. txn

File: src/Storages/MergeTree/MergeTreeDataWriter.cpp (hunk 3 above)
Comment (CheSema) [reply src/Storages/MergeTree/MergeTreeDataWriter.cpp]: removed

**Comment by CheSema** [comment]
It is a fix for the errors like that:

https://d1k2gkhrlfqv31.cloudfront.net/clickhouse-test-reports-private/11353/6087d4032ee83564d7b7b9ac0266c2f19abc015a/integration_tests__asan__old_analyzer__%5B5_6%5D.html

https://d1k2gkhrlfqv31.cloudfront.net/clickhouse-test-reports-private/11353/6087d4032ee83564d7b7b9ac0266c2f19abc015a/integration_tests__asan__old_analyzer__[5_6]//home/ubuntu/actions-runner/_work/_temp/test/output_dir/integration_run_parallel4_0.log

**Comment by CheSema** [comment]
At previous build:
02884_async_insert_skip_settings -- should be fixed now by https://github.com/ClickHouse/ClickHouse/pull/66771

**Review by antonio2368 — APPROVED** [review]

**Comment by CheSema** [comment]
Stateless Tests (msan) [1/4]
03203_client_benchmark_options -- it is new test and it is flak, the fail on the master
https://d1k2gkhrlfqv31.cloudfront.net/clickhouse-test-reports-private/0/51b843e4202728c65e76e3ac5546545febdf8081/stateless_tests__msan__s3_storage__[3_3].html

**Comment by CheSema** [comment]
Stress Test (tsan)
Hung check failed
https://s3.amazonaws.com/clickhouse-test-reports/66279/20f22278bd4bbbf85cff4012fb12c20a03257b13/stress_test__tsan_.html

This is an issue. 
initial query `09b0e14d-2a0a-4530-8620-04319e2b52fa`,
dependent `6222aab4-8ac3-49cf-a9a5-4da98152593f` failed due to exception `CANNOT_SCHEDULE_TASK`
second dependent `e2d5632a-116f-4070-a9fd-2556d734d34d` hanged.

**Comment by CheSema** [comment]
`Integration Tests Flaky Check` all tests has been fixed by Michael Stetsyuk

**Pull request merged**